src/libponyrt/mem/pagemap.c

```diff
@@ -85,31 +88,90 @@ size_t ponyint_pagemap_alloc_size()
 }
 #endif
 
-chunk_t* ponyint_pagemap_get(const void* addr)
+chunk_t* ponyint_pagemap_get_chunk(const void* addr)
 {
   PONY_ATOMIC(void*)* next_node = &root;
   void* node = atomic_load_explicit(next_node, memory_order_acquire);
+  uintptr_t ix = 0;
 
   for(size_t i = 0; i < PAGEMAP_LEVELS; i++)
   {
     if(node == NULL)
       return NULL;
 
-    uintptr_t ix = ((uintptr_t)addr >> level[i].shift) & level[i].mask;
-    next_node = &(((PONY_ATOMIC_RVALUE(void*)*)node)[ix]);
-    node = atomic_load_explicit(next_node, memory_order_acquire);
+    ix = ((uintptr_t)addr >> level[i].shift) & level[i].mask;
+    if (i == PAGEMAP_LEVELS - 1)
+    {
+      // at the lowest level we have to double the index since we store
+      // two pointers (`chunk_t*` and `pony_actor_t*`) instead of one
+      ix = ix * 2;
+      next_node = &(((PONY_ATOMIC_RVALUE(void*)*)node)[ix]);
+      pony_assert(ix <= (uintptr_t)level[i].size);
+      pony_assert(ix/2 <= (uintptr_t)level[i].mask);
+    }
+    else
+    {
+      next_node = &(((PONY_ATOMIC_RVALUE(void*)*)node)[ix]);
+      node = atomic_load_explicit(next_node, memory_order_acquire);
+      pony_assert(ix <= (uintptr_t)level[i].size);
+      pony_assert(ix <= (uintptr_t)level[i].mask);
+    }
+  }
+
+  chunk_t* chunk = (chunk_t*)atomic_load_explicit(next_node, memory_order_acquire);
+  return chunk;
+}
+
+chunk_t* ponyint_pagemap_get(const void* addr, pony_actor_t** actor)
```

Comment:
I think noting why we have a return chunk_t and and out parameter actor rather than two out parameters, would be good documentation for this function.

Reply:
added a comment.. let me know if it's not clear enough..